snap-preload.c

```diff
@@ -12,126 +13,200 @@
 }
 
-
 // Snapd only allows applications to access shared memory paths that match the
 // snap.$SNAP_INSTANCE_NAME.* format. This rewrites paths so that applications
 // don't need changes to conform
-static char *adjust_shm_path(const char *orig_path) {
-  if (!SNAP_INSTANCE_NAME) {
-    return strdup(orig_path);
+int adjust_path(char *new_path, const char *orig_path, size_t len)
+{
+  if (SNAP_INSTANCE_NAME[0] == '\0')
+  {
+    strncpy(new_path, orig_path, len);
   }
-
-  int path_len = strlen(orig_path) + strlen(SNAP_INSTANCE_NAME) + 9;
-  char *new_path = malloc(path_len);
-  assert(new_path);
-  const char *path = (orig_path[0] == '/') ? &(orig_path[1]) : orig_path;
-  snprintf(new_path, path_len , "/snap.%s.%s", SNAP_INSTANCE_NAME, path);
-  log("shm path rewritten: %s -> %s", orig_path, new_path);
-  return new_path;
+  else
+  {
+    if (orig_path[0] == '/')
+      orig_path++;
+    snprintf(new_path, len, "/snap.%s.%s", SNAP_INSTANCE_NAME, orig_path);
+  }
+  return 0;
 }
 
-
 // overrides
 
-int setgroups(size_t size, const gid_t *list) {
+int setgroups(size_t size, const gid_t *list)
+{
   return orig_setgroups(0, NULL);
 }
 
-
 // This is only needed until there's proper support in snapd for initgroups()
 // see https://forum.snapcraft.io/t/seccomp-filtering-for-setgroups/2109 for
 // more info.
-int initgroups(const char *user, gid_t group) {
+int initgroups(const char *user, gid_t group)
+{
   return setgroups(0, NULL);
 }
 
+int shm_open(const char *name, int oflag, mode_t mode)
+{
+  char new_path[NAME_MAX];
+  adjust_path(new_path, name, sizeof(new_path));
+  return orig_shm_open(new_path, oflag, mode);
+}
 
-int shm_open(const char *name, int oflag, mode_t mode) {
-  char *new_path = adjust_shm_path(name);
-  int res = orig_shm_open(new_path, oflag, mode);
-  free(new_path);
-  return res;
+int shm_unlink(const char *name)
+{
+  char new_path[NAME_MAX];
+  adjust_path(new_path, name, sizeof(new_path));
+  return orig_shm_unlink(new_path);
 }
 
+sem_t *sem_open(const char *name, int oflag, ...)
+{
+  va_list args;
+  mode_t mode;
+  unsigned int value;
+  int fd;
+  int existed = 0;
+  sem_t initial_sem;
+  sem_t *final_sem;
+  char new_name[MAX_SEM_NAME_SIZE];
+  char sem_path[NAME_MAX];
+  char tmp_path[NAME_MAX];
+
+  adjust_path(new_name, name, sizeof(new_name));
+  log("new_name %s", new_name);
+
+  if (!(oflag & O_CREAT))
+    return orig_sem_open(new_name, oflag);
+
+  va_start(args, oflag);
+  mode = va_arg(args, mode_t);
+  value = va_arg(args, unsigned int);
+  va_end(args);
+  if (value > SEM_VALUE_MAX)
+  {
+    errno = EINVAL;
+    return SEM_FAILED;
+  }
 
-int shm_unlink(const char *name) {
-  char *new_path = adjust_shm_path(name);
-  int res = orig_shm_unlink(new_path);
-  free(new_path);
-  return res;
+  // glibc creates a tempfile, initializes it, hardlink to the requested name
+  // and finally unlinks the tempfile. Due to snap confinement rules, the first
+  // step fails in the original implementation.
```

Comment:
Which original implementation? Possibly add a link to the bug here?

Reply:
the libc code